src/v/coding-style.md

```diff
@@ -6,12 +6,50 @@ The document covers style and convention not enforced by clang-format.
 
 Header files have the `.h` extension, source files use the `.cc` extension.
 
+Public header files should go in an `includes` directory within the module they belong to. 
+For example, a module named `foo` should have public header files within `src/v/foo/include/foo`.
+
 Test files go in the `tests` directory of the module they belong to. That is,
 the tests for a feature named `foo` and located in `src/v/foo` will be located
 in `src/v/foo/tests`.
 
 Use `snake_case` naming convention by default for files and language identifiers.
 
+## Documentation
+
+### Modules
+
+Modules themselves should have a `README.md` file explaining the purpose of the module,
+it's API surface, and an architectural overview of that module (mentioning major classes),
+etc. This will aid people new to the module in understanding the lay of the land. When in doubt
+of if information should be placed in a class or in a README, prefer moving the documentation
+as close to the source as possible (inside the class documentation). We want to prioritize
+documentation not becoming stale - and also not cognitively overload the reader of the module
+README.
+
+### Classes
+
+Classes in public headers should be documented, plain structs that only hold data are not
+required to be documented, but it is often helpful to add. Writing points when documenting
+a given class:
+
+* Lifecycle of the class
+* Place within the larger module
+* How to use it correctly
+* Broad strokes of implmentation details
+* Which cores it exists on (if a sharded service)
+* If it is a cluster/node/shard view of information
+
+While only required in public headers, it's also encouraged to add this information to key classes 
+within a given module.
+
+### Functions
+
+Non trivial free and class functions in public headers should be documented. Trivial is ultimately
+a judgement call, but a rule of thumb is that getter methods (like `iobuf value() const { return _value; }`)
+or other common lifecycle methods like `start` and `stop` could be considered trivial. It's highly
+recommended to error on the side of verbosity when documenting functions however.
```

Comment:
I wonder if we can find (1) some article about how to write good function documentation and (2) a code base with good examples. Sometimes I catch myself writing prose that matches 1-1 with the code, but I suspect that isn't really the type of doc string we're looking for.

Reply:
Good call. Let me look around a bit and get back to this

Reply:
I think cockroach is a decent example.

They have "package" level documentation (which we'll solve with README): https://github.com/cockroachdb/cockroach/blob/master/pkg/gossip/doc.go

If you look through the entry point to that exact package, I think it's pretty solid:
https://github.com/cockroachdb/cockroach/blob/master/pkg/gossip/gossip.go

Another package example from cockroach:
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/doc.go, or https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlstats/ssprovider.go

Anyways, I don't think more text is better, but something is better than nothing. Maybe it's better to surface stuff in our codebase?

Example of something good IMO:
https://github.com/redpanda-data/redpanda/blob/697d4a118e45aa282fcd0d5d7daef47b30156558/src/v/raft/state_machine_manager.h#L65

Example of something that could be better IMO:
https://github.com/redpanda-data/redpanda/blob/697d4a118e45aa282fcd0d5d7daef47b30156558/src/v/cluster/controller.h#L39

Reply:
This looks great to me!